include/stl2/detail/iterator/any_iterator.hpp

```diff
@@ -0,0 +1,230 @@
+            }
+            InputIterator{I} cursor(I i, big_tag) {
+                data_.big = new shared_iterator<I>(std::move(i));
+                deref_ = &deref_big<Reference, I>;
+                exec_ = &exec_big<RValueReference, I>;
+            }
+        public:
+            using value_type = ValueType;
+            using single_pass = std::true_type;
+            struct mixin;
+            cursor() = default;
+            cursor(cursor &&that) {
+                that.exec_(op::move, &that.data_, &data_);
+                swap(deref_, that.deref_);
+                swap(exec_, that.exec_);
+            }
+            cursor(cursor const &that) {
+                that.exec_(op::copy, const_cast<blob *>(&that.data_), &data_);
+                deref_ = that.deref_;
+                exec_ = that.exec_;
+            }
+            cursor &operator=(cursor &&that) {
+                if (&that != this) {
+                    this->~cursor();
+                    ::new (static_cast<void *>(this))
+                        cursor(std::move(that));
+                }
+                return *this;
+            }
+            cursor &operator=(cursor const &that) {
+                if (&that != this) {
+                    this->~cursor();
+                    ::new (static_cast<void *>(this)) cursor(that);
+                }
+                return *this;
+            }
+            ~cursor() {
+                exec_(op::nuke, &data_, nullptr);
+            }
+            Reference read() const {
+                return deref_(data_);
+            }
+            bool equal(cursor const &that) const {
+                return exec_(op::comp, const_cast<blob *>(&data_),
+                    const_cast<blob *>(&that.data_)) != nullptr;
+            }
+            void next() {
+                exec_(op::bump, &data_, nullptr);
+            }
+            RValueReference indirect_move() const {
+                return exec_(op::rval, nullptr, nullptr)(data_);
+            }
+        };
+
+        template<class Reference, class ValueType, class RValueReference>
+        struct cursor<Reference, ValueType, RValueReference>::mixin
+        : basic_mixin<cursor> {
+            mixin() = default;
+            using mixin::basic_mixin::basic_mixin;
+            InputIterator{I} mixin(I i)
+            requires
+                ConvertibleTo<reference_t<I>, Reference>() &&
+                ConvertibleTo<rvalue_reference_t<I>, RValueReference>()
+            : basic_mixin<cursor>(cursor{std::move(i), is_small<I>{}}) {
+            }
+            template <class I, class C = cursor,
+                class = meta::if_c<!is_same<I, mixin>::value>>
+            requires
+                !Same<I, basic_iterator<C>>() && InputIterator<I>() &&
+                ConvertibleTo<reference_t<I>, Reference>() &&
+                ConvertibleTo<rvalue_reference_t<I>, RValueReference>()
+            mixin &operator=(I i) & {
```

Comment:
If the goal is only to support reasonable semantics, then I don't think we need a "protocol" hook for this in `mixin`; we should have an assignment operator overload in `basic_iterator` that simply forwards assignments to the cursor. This encourages slightly tighter semantics and at the same time makes it easier to assign to the cursor:

```diff
diff --git a/include/stl2/detail/iterator/any_iterator.hpp b/include/stl2/detail/iterator/any_iterator.hpp
index a4ae014..97f804a 100644
--- a/include/stl2/detail/iterator/any_iterator.hpp
+++ b/include/stl2/detail/iterator/any_iterator.hpp
@@ -159,6 +159,9 @@ STL2_OPEN_NAMESPACE {
                 deref_ = &deref_big<Reference, I>;
                 exec_ = &exec_big<RValueReference, I>;
             }
+            InputIterator{I} cursor(I i)
+            : cursor{i, is_small<I>{}}
+            {}
             void reset() noexcept {
                 exec_(op::nuke, &data_, nullptr);
                 deref_ = &uninit_deref<Reference>;
@@ -179,7 +182,6 @@ STL2_OPEN_NAMESPACE {
         public:
             using value_type = ValueType;
             using single_pass = std::true_type;
-            struct mixin;
             cursor() = default;
             cursor(cursor &&that) {
                 move_from(that);
@@ -218,30 +220,7 @@ STL2_OPEN_NAMESPACE {
                 return exec_(op::rval, nullptr, nullptr)(data_);
             }
         };
-
-        template<class Reference, class ValueType, class RValueReference>
-        struct cursor<Reference, ValueType, RValueReference>::mixin
-        : basic_mixin<cursor> {
-            mixin() = default;
-            using mixin::basic_mixin::basic_mixin;
-            InputIterator{I} mixin(I i)
-            requires
-                ConvertibleTo<reference_t<I>, Reference>() &&
-                ConvertibleTo<rvalue_reference_t<I>, RValueReference>()
-            : basic_mixin<cursor>(cursor{std::move(i), is_small<I>{}}) {
-            }
-            template <class I, class C = cursor,
-                class = meta::if_c<!is_same<I, mixin>::value>>
-            requires
-                !Same<I, basic_iterator<C>>() && InputIterator<I>() &&
-                ConvertibleTo<reference_t<I>, Reference>() &&
-                ConvertibleTo<rvalue_reference_t<I>, RValueReference>()
-            mixin &operator=(I i) & {
-                this->get() = cursor{i, is_small<I>{}};
-                return *this;
-            }
-        };
-	}
+    }
 
     template<class Reference,
         class ValueType = __uncvref<Reference>,
diff --git a/include/stl2/detail/iterator/basic_iterator.hpp b/include/stl2/detail/iterator/basic_iterator.hpp
index c16067d..13c25d8 100644
--- a/include/stl2/detail/iterator/basic_iterator.hpp
+++ b/include/stl2/detail/iterator/basic_iterator.hpp
@@ -602,12 +602,12 @@ STL2_OPEN_NAMESPACE {
 		// http://wg21.link/P0186
 		template <class O>
 		requires
-			__assignable<mixin&, O> && !models::Same<decay_t<O>, mixin> &&
-			!models::Same<decay_t<O>, basic_iterator>
+			!models::Same<decay_t<O>, basic_iterator> &&
+			models::Assignable<C&, O>
 		constexpr basic_iterator& operator=(O&& o) &
-		noexcept(is_nothrow_assignable<mixin&, O>::value)
+		noexcept(is_nothrow_assignable<C&, O>::value)
 		{
-			static_cast<mixin&>(*this) = forward<O>(o);
+			get() = __stl2::forward<O>(o);
 			return *this;
 		}
```